spring-cloud-gcp-starters/spring-cloud-gcp-starter-core/src/main/java/org/springframework/cloud/gcp/core/autoconfig/GcpContextAutoConfiguration.java

```diff
@@ -54,4 +59,15 @@ public GoogleCredentials googleCredentials() throws Exception {
 		}
 	}
 
+	/**
+	 * @return a {@link GcpProjectIdProvider} that returns the project ID in the properties or, if
+	 * none, the project ID from the GOOGLE_CLOUD_PROJECT envvar and Metadata Server
+	 */
+	@Bean
+	@ConditionalOnMissingBean
+	public GcpProjectIdProvider gcpProjectIdProvider() {
+		return () -> gcpProperties.getProjectId() != null
+				? Optional.of(gcpProperties.getProjectId())
+				: Optional.ofNullable(ServiceOptions.getDefaultProjectId());
```

Comment:
I'm now thinking whether the method above should also return a CredentialsProvider instead of GoogleCredentials. It would make unit testing easier, and design would be arguably better.